File: vector/vector.go
package vector

import "math"

type Vector struct {
	X, Y, Z float64
}

func (v Vector) Add(r Vector) Vector {
	return Vector{v.X + r.X, v.Y + r.Y, v.Z + r.Z}
}

func (v Vector) Scale(r float64) Vector {
	return Vector{v.X * r, v.Y * r, v.Z * r}
}

func (v Vector) DotProduct(r Vector) float64 {
	return v.X*r.X + v.Y*r.Y + v.Z*r.Z
}

func (v Vector) CrossProduct(r Vector) Vector {
	return Vector{v.Y*r.Z - v.Z*r.Y, v.Z*r.X - v.X*r.Z, v.X*r.Y - v.Y*r.X}
}

func (v Vector) Normalize() Vector {
	return v.Scale(1.0 / math.Sqrt(v.DotProduct(v)))
}

func New(x, y, z float64) Vector {
	return Vector{x, y, z}
}

File: rustrays/vector.rs
use std::to_str::ToStr;
use std::num::rsqrt;

pub struct Vector {
	x: f64,
	y: f64,
	z: f64
}

impl Vector {
	pub fn add(&self, r: &Vector) -> Vector {
		Vector { x: r.x + self.x, y: r.y + self.y, z: r.z + self.z }
	}

	pub fn scale(&self, r: f64) -> Vector {
		Vector { x: self.x * r, y: self.y * r, z: self.z * r }
	}

	pub fn dotProduct(&self, r: &Vector) -> f64 {
		return self.x*r.x + self.y*r.y + self.z*r.z
	}

	pub fn crossProduct(&self, r: Vector) -> Vector {
		Vector { x: self.y * r.z - self.z * r.y, y: self.z * r.x - self.x * r.z, z: self.x * r.y - self.y * r.x }
	}

	pub fn normalize(&self) -> Vector {
		self.scale(rsqrt(self.x * self.x + self.y * self.y + self.z * self.z))
	}
}

impl ToStr for Vector {
	fn to_str(&self) -> ~str {
		format!("Vector [x={}, y={}, z={}]", self.x, self.y, self.z)
	}
}
